metagpt/provider/openai_api.py

```diff
@@ -127,10 +116,20 @@ def get_costs(self) -> Costs:
         return Costs(self.total_prompt_tokens, self.total_completion_tokens, self.total_cost, self.total_budget)
 
 
+def log_and_reraise(retry_state):
+    logger.error(f"Retry attempts exhausted. Last exception: {retry_state.outcome.exception()}")
```

Comment:
If the `exception()` return all Traceback messages or just the last one. If it's better to use `logger.exception` to catch the `exceptions()`?

Reply:
> If the `exception()` return all Traceback messages or just the last one. If it's better to use `logger.exception` to catch the `exceptions()`?

I prefer giving *recommend log text*, like this:
<img width="1977" alt="image" src="https://github.com/geekan/MetaGPT/assets/12390909/c788ce7f-7abd-448b-bf78-1f3a6fdce96a">

By the way, I am interest to test what you said, but getting a object print when I `print(logger.exception)`

Reply:
the usage is `logger.exception(f"Retry attempts exhausted. Last exception: {retry_state.outcome.exception()}")`. You can have a test if it will print the whole Traceback like below in the log file
![image](https://github.com/geekan/MetaGPT/assets/7869691/7afb3871-c097-4310-b48d-182d9f622ee4)

Reply:
> the usage is `logger.exception(f"Retry attempts exhausted. Last exception: {retry_state.outcome.exception()}")`. You can have a test if it will print the whole Traceback like below in the log file ![image](https://user-images.githubusercontent.com/7869691/259103333-7afb3871-c097-4310-b48d-182d9f622ee4.png)

Thanks for your advicing, and I am satisfied to learn to use it. 
But as my further test, `logger.exception` has no trace stack at that time. Let's see:
<img width="2185" alt="image" src="https://github.com/geekan/MetaGPT/assets/12390909/4f669c83-88ac-44a5-b6b9-1c9961a77969">